cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,426 @@
+      n_edges_(other.n_edges_),
+      edges_(other.edges_),
+      vertices_(other.vertices_),
+      complement_(other.complement_)
+  {
+  }
+
+  ~graph_mask_view_t()                            = default;
+  graph_mask_view_t(graph_mask_view_t&&) noexcept = default;
+
+  graph_mask_view_t& operator=(graph_mask_view_t&&) noexcept = default;
+  graph_mask_view_t& operator=(graph_mask_view_t const& other) = default;
+
+  /**
+   * Return whether or not a specific vertex is masked
+   * @param vertex
+   * @return
+   */
+  __device__ bool is_vertex_masked(vertex_t vertex)
+  {
+    if (has_vertex_mask_) {
+      // TODO: This is going to implicitly limit vertex_t to uint32_t. Need to support 64-bit types
+      // as well
+      return detail::_get_bit<mask_t>(vertices_, static_cast<mask_t>(vertex));
+    } else {
+      return !complement_;
+    }
+  }
+
+  /**
+   * Return whether or not a specific edge is masked
+   * @param edge_offset
+   * @return
+   */
+  __device__ bool is_edge_masked(edge_t edge_offset)
+  {
+    if (has_edge_mask_) {
+      // TODO: This is going to implicitly limit edge_t to uint32_t. Need to support 64-bit types as
+      // well
+      return detail::_get_bit<mask_t>(edges_, static_cast<mask_t>(edge_offset));
+    } else {
+      return !complement_;
+    }
+  }
+
+  /**
+   * Add specific vertex to mask
+   * @param vertex id of vertex to mask
+   * @return
+   */
+  __device__ void mask_vertex(vertex_t vertex)
+  {
+    RAFT_EXPECTS(has_vertex_mask_, "Vertex mask needs to be initialized before it can be used");
+    detail::_set_bit<mask_t>(vertices_, static_cast<mask_t>(vertex), true);
+  }
+
+  /**
+   * Add specific edge to mask
+   * @param edge_offset offset of edge to mask
+   * @return
+   */
+  __device__ void mask_edge(edge_t edge_offset)
+  {
+    RAFT_EXPECTS(has_edge_mask_, "Edge mask needs to be initialized before it can be used");
+    detail::_set_bit<mask_t>(edges_, static_cast<mask_t>(edge_offset), true);
+  }
+
+  /**
+   * Are masks complemeneted?
+   *
+   * - !complemented means masks are inclusive (masking in)
+   * - complemented means masks are exclusive (masking out)
+   */
+  __host__ __device__ bool is_complemented() const { return complement_; }
+
+  /**
+   * Has the edge mask been initialized?
+   */
+  __host__ __device__ bool has_edge_mask() const { return has_vertex_mask_; }
+
+  /**
+   * Has the vertex mask been initialized?
+   */
+  __host__ __device__ bool has_vertex_mask() const { return has_edge_mask_; }
+
+  /**
+   * Get the vertex mask
+   */
+  __host__ __device__ mask_t* get_vertex_mask() { return vertices_; }
+
+  /**
+   * Get the edge mask
+   */
+  __host__ __device__ mask_t* get_edge_mask() { return edges_; }
+
+ protected:
+  raft::handle_t const& handle_;
+  bool has_vertex_mask_{false};
+  bool has_edge_mask_{false};
+  vertex_t n_vertices_;
+  edge_t n_edges_;
+  bool complement_{false};
+  mask_t* vertices_;
+  mask_t* edges_;
```

Comment:
I am not having a single mask_t array will work for edges in multi-GPU. Each GPU holds multiple edge partitions, and shouldn't we better have one array per edge partition?